compiler/modules/front/const.bal

```diff
@@ -517,47 +520,88 @@ function foldVarRefExpr(FoldContext cx, t:SemType? expectedType, s:VarRefExpr ex
 }
 
 function foldIntLiteralExpr(FoldContext cx, t:SemType? expectedType, s:IntLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    float|int|error result;
-    string ty;
-    if expr.base == 10 && expectsFloat(expectedType) {
-        result = floatFromDecimalLiteral(expr.digits);
-        ty = "float"; 
+    int|float|decimal result;
+    t:Context tx = cx.typeContext();
+    Position startPos = expr.startPos;
+    if expectedType == () || expectsInt(tx, expectedType) {
+        result = check intFromLiteral(cx, expr);
+    }  
+    else if expectsFloat(tx, expectedType) {
+        result = floatFromFpLiteral(expr.digits);
+    }
+    else if expectsDecimal(tx, expectedType) {
+        result = check decimalFromFpLiteral(cx, expr.digits, startPos);
     }
     else {
-        result = s:intFromIntLiteral(expr.base, expr.digits);
-        ty = "int";
-    }
-    if result is int|float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
-    }
-    else {
-        return cx.semanticErr("invalid " + ty + " literal", cause=result, pos=expr.startPos);
+        result = check intFromLiteral(cx, expr);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
+}
+
+function expectsInt(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:INT));
+}
+
+function expectsFloat(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:FLOAT));
+}
+
+function expectsDecimal(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:DECIMAL));
 }
 
-function expectsFloat(t:SemType? semType) returns boolean {
-    if semType == () {
-        return false;
+// Since the binary floating point literal is parsed correctly,
+// it is impossible to throw an error.
+function floatFromFpLiteral(string digits) returns float {
+    float|error f = float:fromString(digits);
+    if f is error {
+        panic err:impossible();
     }
-    else {
-        return t:isSubtypeSimple(t:intersect(semType, t:union(t:FLOAT, t:INT)), t:FLOAT);
+    return f;
+}
+
+// Even if the decimal floating point literal is parsed correctly,
+// infinities should through an error.
+function decimalFromFpLiteral(FoldContext cx, string decimalStr, Position pos) returns decimal|FoldError {
+    decimal|error d = decimal:fromString(decimalStr);
+    if d is error {
+        return cx.semanticErr("invalid decimal literal", pos);
     }
+    return d;
 }
 
-function floatFromDecimalLiteral(string digits) returns float|error {
-    return float:fromString(digits);
+// Even if the integer literal is parsed correctly,
+// infinities should through an error.
+function intFromLiteral(FoldContext cx, s:IntLiteralExpr expr) returns int|FoldError {
+    int|error i = s:intFromIntLiteral(expr.base, expr.digits);
+    if i is error {
+        return cx.semanticErr("invalid int literal", expr.startPos);
```

Comment:
For what input will this happen?

Reply:
For values greater than int max or less than min (ex: 9223372036854775808).